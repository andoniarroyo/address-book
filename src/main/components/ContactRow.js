// @flow
// external modules
import React from 'react';
import { Link } from 'react-router-dom';
import countriesFactory from 'country-list';
// internal types
import type { ContactType } from '../index';
import type { RemoveContactActionFactoryType } from '../actions';

const countries = countriesFactory();

export type ContactRowPropsType = {
  contact: ContactType,
  removeContact: RemoveContactActionFactoryType,
};

/**
 * The contact render method
 * @param {ContactRowPropsType} props The properties to render the object
 * @returns {React$Element<any>} HTML markup for the component
 */
const ContactRow = ({ contact, removeContact }: ContactRowPropsType) => {
  const emailLink = `mailto:${contact.email}`;
  return (
    <article className="contact">
      <span className="item">
        <Link to={`/edit/${contact.id}`}>
          {contact.firstName}
        </Link>
      </span>
      <span className="item">{contact.lastName}</span>
      <span className="item">
        <a href={emailLink}>
          {contact.email}
        </a>
      </span>
      <span className="item">{countries.getName(contact.countryCode)}</span>
      <span className="item item-final">
        <button id="remove-button" onClick={() => removeContact(contact.id)}>
          Remove it!
        </button>
      </span>
    </article>
  );
};

export default ContactRow;
