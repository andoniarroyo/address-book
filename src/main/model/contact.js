// @flow
// external modules
import shortid from 'shortid';
// internal types
import type { ContactPartialType, ContactType } from '../index';
// internal modules
import { DEFAULT_COUNTRY_CODE } from '../constants';

/**
 * Checks if the incoming string exists and is longer than 0
 * @param {string} stringToValidate The string to be evaluated
 * @return {boolean} True if the string is longer than 0, False otherwise
 * @private
 */
const existingString = (stringToValidate: string): boolean => !!(stringToValidate);

/**
 * Checks if the email reported is a valid email
 * @param {string} emailToValidate Email account to be evaluated
 * @return {boolean} True if the email is valid, False otherwise
 * @private
 */
const isValidEmail = (emailToValidate: string): boolean => (
  new RegExp(/(\w+)@(\w+).[a-zA-Z]/g).test(emailToValidate)
);

/**
 * Checks if the incoming contact has a valid values for the required properties
 * @param {ContactType} contact The contact to be evaluated
 * @return {boolean} True if the contact is valid, False otherwise
 * @public
 */
export const isValidContact = (contact: ContactType): boolean => (
  Boolean(contact && existingString(contact.id) && existingString(contact.firstName)
    && existingString(contact.lastName) && isValidEmail(contact.email) && existingString(contact.countryCode))
);

/**
 * Creates a new contact with the default values for the non informed properties
 * @param {ContactPartialType} contactPartial Contact partially informed to be used as refeence
*  @return {ContactType} The created contact
 * @public
 */
export const createContact = ({
  id = shortid.generate(), firstName = '', lastName = '', email = '', countryCode = DEFAULT_COUNTRY_CODE,
}
: ContactPartialType): ContactType => (
  {
    id, firstName, lastName, email, countryCode,
  }
);
