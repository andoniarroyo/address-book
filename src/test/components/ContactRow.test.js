// external modules
import React from 'react';
import renderer from 'react-test-renderer';
import { mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
// internal modules
import ContactRow from '../../main/components/ContactRow';
import { createContact } from '../../main/model/contact';
import staticRouterWrapper from '../utils/staticRouterWrapper';

describe('<ContactRow />', () => {
  beforeAll(() => {
    configure({ adapter: new Adapter() });
  });
  const createProps = () => (
    {
      contact: createContact({
        id: '1', firstName: 'Andoni', lastName: 'Arroyo', email: 'andoni.arroyo@gmail.com', countryCode: 'GB',
      }),
      removeContact: jest.fn(),
    }
  );
  describe('checking the layout', () => {
    it('renders the component', () => {
      const props = createProps();
      const tree = renderer.create(staticRouterWrapper(<ContactRow {...props} />)).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('checking the behaviour', () => {
    it('ask to remove the contact when the remove button is clicked', () => {
      const props = createProps();
      const wrapper = mount(staticRouterWrapper(<ContactRow {...props} />));
      wrapper.find('#remove-button').simulate('click');
      expect(props.removeContact).toBeCalledWith(props.contact.id);
    });
  });
});
