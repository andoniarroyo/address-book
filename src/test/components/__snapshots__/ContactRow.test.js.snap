// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactRow /> checking the layout renders the component 1`] = `
<article
  className="contact"
>
  <span
    className="item"
  >
    <a
      href="/edit/1"
      onClick={[Function]}
    >
      Andoni
    </a>
  </span>
  <span
    className="item"
  >
    Arroyo
  </span>
  <span
    className="item"
  >
    <a
      href="mailto:andoni.arroyo@gmail.com"
    >
      andoni.arroyo@gmail.com
    </a>
  </span>
  <span
    className="item"
  >
    United Kingdom
  </span>
  <span
    className="item item-final"
  >
    <button
      id="remove-button"
      onClick={[Function]}
    >
      Remove it!
    </button>
  </span>
</article>
`;
